{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fd7f82e8-96bd-48b4-80aa-92b0578492a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.12.0\n"
     ]
    }
   ],
   "source": [
    "from platform import python_version\n",
    "print(python_version())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "693974f9-d0e9-43dd-9256-5b70c1a19a2e",
   "metadata": {},
   "source": [
    "## Gemini API\n",
    "\n",
    "https://ai.google.dev/gemini-api/docs\n",
    "\n",
    "#### API key - Free of charge\n",
    "\n",
    "https://aistudio.google.com/app/apikey\n",
    "\n",
    "AIzaSyA1ZXcSe6NP5jiIw93sUpZYb8RKK1PgYDE\n",
    "\n",
    "### Google Enable API\n",
    "\n",
    "  - You are about to enable 'Generative Language API'.\n",
    "\n",
    "https://ai.google.dev/gemini-api/docs/oauth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d89870d3-aff7-49aa-81ba-036cc27e04cf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "fd0fc3a5-3f5a-4bd9-8d67-84ef0275adf3",
   "metadata": {},
   "source": [
    "### Google Python projects\n",
    "\n",
    "#### Gemini API Text Implementation\n",
    "\n",
    "https://github.com/RepellentSpy/Gemini-API-Text-Implementation/tree/main\n",
    "\n",
    "#### gemini-api 0.1.6\n",
    "\n",
    "https://pypi.org/project/gemini-api/\n",
    "\n",
    "\n",
    "#### Gemini-API\n",
    "\n",
    "https://github.com/dsdanielpark/Gemini-API"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2069d3bb-c1e2-4598-9156-a517ad2e1c2b",
   "metadata": {},
   "source": [
    "## LLM - Large Language Model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8835a112-df87-494e-9882-68f8246af45b",
   "metadata": {},
   "source": [
    "### Gemini flash\n",
    "\n",
    "gemini-1.5-flash-latest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f086b501-7252-46e6-82f1-ebb403319bc4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import torch\n",
    "torch.cuda.is_available()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1ca8fe6a-05f5-404c-b1f6-45cdd4148943",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !nvidia-smi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0a82f735-9f90-4bb2-bdf2-79a927162f6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !lsmod | grep nvidia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "24db0cf4-e681-493b-9208-f832bdf0c175",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>:root { --jp-notebook-max-width: 100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'root_chibe': '../../chibe/', 'root_colab': '../../colaboracoes/', 'root0': '../../colaboracoes/covid/sonia_andrade/taubate/proteomics_202205/', 'project': 'Taubate COVID-19', 's_project': 'taubate_covid19', 'gene_protein': 'protein', 's_omics': 'proteomics', 'has_age': True, 'has_gender': True, 'want_normalized': False, 'abs_lfc_cutoff_inf': 0.4, 's_pathw_enrichm_method': 'enricher', 'num_min_degs_for_ptw_enr': 3, 'tolerance_pathway_index': 0.15, 'case_list': ['g2a_male', 'g2a_female', 'g2b_male', 'g2b_female', 'g3_male_adult', 'g3_male_elder', 'g3_female_adult', 'g3_female_elder'], 'pval_pathway_cutoff': 0.05, 'fdr_pathway_cutoff': 0.05, 'num_of_genes_cutoff': 3, 'API_KEY': 'AIzaSyA1ZXcSe6NP5jiIw93sUpZYb8RKK1PgYDE', 'disease': 'COVID-19', 'context_disease': 'COVID-19 is the disease caused by the virus SARS-CoV-2. When the virus enters your body, it infects cells, primarily in the respiratory system.', 'n_sentences': 5, 'run_list': ['run01', 'run02'], 'chosen_model_list': [1, 3], 'i_dfp_list': [0, 1, 2, 3], 'chosen_model_sampling': 3, 'terms1_param': ['OR', 'COVID', 'SARS-CoV-2'], 'terms2_param': [], 'terms_not_param': ['NOT', 'MERS', 'SARS-CoV-1'], 'connective_param': 'AND', 'remove_synonym_list': ['CAP', 'MV', 'MDB'], 'inidate': '2019/10/01', 'enddate': '2030/12/31', 'verbose_query': False, 'force_query': False, 'sleep_entrez': [5, 7, 10], 'retmax': 100000, 'try_all_text': True, 'text_quote': '', 'dec_ncpus': 2, 'sleep_TIKA': [10, 20, 30], 'min_words_text': 100}\n"
     ]
    }
   ],
   "source": [
    "import os, sys, pickle\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.set_option('display.width', 100)\n",
    "pd.set_option('max_colwidth', 80)\n",
    "import yaml\n",
    "\n",
    "import seaborn as sns\n",
    "sns.set_context(\"notebook\", font_scale=1.4)\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "%matplotlib inline\n",
    "\n",
    "sys.path.insert(1, '../src/')\n",
    "\n",
    "from Basic import *\n",
    "from entrez_conversion import *\n",
    "from pubmed_lib import *\n",
    "from biopax_lib import *\n",
    "from gemini_lib import *\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "from IPython.display import display, HTML\n",
    "# display(HTML(\"<style>.container { width:100% !important; }</style>\"))\n",
    "display(HTML(\"<style>:root { --jp-notebook-max-width: 100% !important; }</style>\"))\n",
    "\n",
    "email = \"flalix@gmail.com\"\n",
    "\n",
    "# !pip3 install pyyaml\n",
    "with open('config_taubate.yml', 'r') as file:\n",
    "    dic_yml = yaml.safe_load(file)\n",
    "\n",
    "print(dic_yml)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "db0e17e2-3775-4709-9873-2f918e1ed275",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "G/P LFC cutoffs: lfc=1.000; fdr=0.050\n",
      "Pathway cutoffs: pval=0.050; fdr=0.050; num of genes=3\n"
     ]
    }
   ],
   "source": [
    "root_chibe = dic_yml['root_chibe']\n",
    "root_colab = dic_yml['root_colab']\n",
    "root0 = dic_yml['root0']\n",
    "\n",
    "project = dic_yml['project']\n",
    "s_project = dic_yml['s_project']\n",
    "\n",
    "gene_protein = dic_yml['gene_protein']\n",
    "s_omics = dic_yml['s_omics']\n",
    "\n",
    "has_age = dic_yml['has_age']\n",
    "has_gender = dic_yml['has_gender']\n",
    "\n",
    "want_normalized = dic_yml['want_normalized']\n",
    "\n",
    "abs_lfc_cutoff_inf = dic_yml['abs_lfc_cutoff_inf']\n",
    "s_pathw_enrichm_method = dic_yml['s_pathw_enrichm_method']\n",
    "num_min_degs_for_ptw_enr = dic_yml['num_min_degs_for_ptw_enr']\n",
    "tolerance_pathway_index = dic_yml['tolerance_pathway_index']\n",
    "\n",
    "case_list = dic_yml['case_list']\n",
    "\n",
    "pval_pathway_cutoff = dic_yml['pval_pathway_cutoff']\n",
    "fdr_pathway_cutoff = dic_yml['fdr_pathway_cutoff']\n",
    "num_of_genes_cutoff = dic_yml['num_of_genes_cutoff']\n",
    "\n",
    "run_list = dic_yml['run_list']\n",
    "chosen_model_list = dic_yml['chosen_model_list']\n",
    "i_dfp_list = dic_yml['i_dfp_list']\n",
    "\n",
    "exp_normalization='quantile_norm' if want_normalized else None\n",
    "normalization='not_normalized' if exp_normalization is None else exp_normalization\n",
    "\n",
    "cfg = Config(project, s_project, case_list, root0)\n",
    "\n",
    "case = case_list[0]\n",
    "\n",
    "n_genes_annot_ptw, n_degs, n_degs_in_ptw, n_degs_not_in_ptw, degs_in_all_ratio = -1,-1,-1,-1,-1\n",
    "abs_lfc_cutoff, fdr_lfc_cutoff, n_degs, n_degs_up, n_degs_dw = cfg.get_best_lfc_cutoff(case, 'not_normalized')\n",
    "\n",
    "print(f\"G/P LFC cutoffs: lfc={abs_lfc_cutoff:.3f}; fdr={fdr_lfc_cutoff:.3f}\")\n",
    "print(f\"Pathway cutoffs: pval={pval_pathway_cutoff:.3f}; fdr={fdr_pathway_cutoff:.3f}; num of genes={num_of_genes_cutoff}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "75d9a3d5-6a54-4b70-9dd2-deb3ddb09810",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start opening tables ....\n",
      "Building synonym dictionary ...\n",
      "\n",
      "\n",
      "Echo Parameters:\n",
      "For case 'g3_male_elder', there are 140/140 DAPs/DAPs with ensembl_id\n",
      "DAP's cutoffs: abs(LFC)=0.600; FDR=0.400\n",
      "\t140/140 DAPs/ensembl.\n",
      "\t\tUp 40/40 DAPs/ensembl.\n",
      "\t\tDw 100/100 DAPs/ensembl.\n",
      "\n",
      "Found 60 (best=60) pathways for geneset num=0 'Reactome_2022'\n",
      "Pathway cutoffs p-value=0.050 fdr=0.050 min genes=3\n",
      "DAPs found in enriched pathways:\n",
      "\tThere are 140 DAPs found in pathways\n",
      "\t105 (best=105) DAPs in pathways and 35/35 DAPs/ensembl not in pathways\n",
      "\n",
      "\t34 DAPs ensembl Up in pathways\n",
      "\t6 DAPs Up ensembl not in pathways\n",
      "\n",
      "\t71 DAPs ensembl Dw in pathways\n",
      "\t29 DAPs Dw ensembl not in pathways\n"
     ]
    }
   ],
   "source": [
    "bpx = Biopax(gene_protein, s_omics, project, s_project, root0,\n",
    "             case_list, has_age, has_gender, clone_objects=False,\n",
    "             exp_normalization=exp_normalization, geneset_num=0, \n",
    "             num_min_degs_for_ptw_enr=num_min_degs_for_ptw_enr, \n",
    "             tolerance_pathway_index=tolerance_pathway_index, \n",
    "             s_pathw_enrichm_method = s_pathw_enrichm_method)\n",
    "\n",
    "case = case_list[5]\n",
    "\n",
    "bpx.cfg.set_default_best_lfc_cutoff(normalization, abs_lfc_cutoff=1, fdr_lfc_cutoff=0.05)\n",
    "ret, degs, degs_ensembl, dfdegs = bpx.open_case(case, verbose=False)\n",
    "print(\"\\nEcho Parameters:\")\n",
    "bpx.echo_parameters()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dacf3997-f558-4b3d-878b-4538b05b5d61",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('enricher_Reactome_2022_taubate_covid19_proteomics_for_g3_male_elder_x_ctrl_not_normalized_cutoff_lfc_0.600_fdr_0.400.tsv',\n",
       " 'enricher_Reactome_2022_taubate_covid19_proteomics_for_g3_male_elder_x_ctrl_not_normalized_cutoff_lfc_0.600_fdr_0.400_pathway_pval_0.050_fdr_0.050_num_genes_3.tsv')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fname, fname_cutoff=bpx.set_enrichment_name()\n",
    "fname, fname_cutoff"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e5ef1c6-da9c-40c5-8e8c-651600c3b974",
   "metadata": {},
   "source": [
    "### Run all, after finding a case\n",
    "  - please open dfr (reactome) before"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "39e06289-0bce-414c-aa60-a707526fc8e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "60"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Nenr=len(bpx.df_enr)\n",
    "Nenr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "cb93f0de-13aa-4a71-804c-5f198f9d7997",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway</th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pval</th>\n",
       "      <th>fdr</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>genes</th>\n",
       "      <th>num_of_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Platelet Degranulation</td>\n",
       "      <td>R-HSA-114608</td>\n",
       "      <td>4.890639e-33</td>\n",
       "      <td>2.239913e-30</td>\n",
       "      <td>48.182590</td>\n",
       "      <td>3584.687599</td>\n",
       "      <td>['LGALS3BP', 'SERPINA3', 'ECM1', 'ORM1', 'ITIH3', 'SERPINA1', 'F13A1', 'LOC1...</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Response To Elevated Platelet Cytosolic Ca2+</td>\n",
       "      <td>R-HSA-76005</td>\n",
       "      <td>1.555951e-32</td>\n",
       "      <td>3.563128e-30</td>\n",
       "      <td>45.832030</td>\n",
       "      <td>3356.767042</td>\n",
       "      <td>['LGALS3BP', 'SERPINA3', 'ECM1', 'ORM1', 'ITIH3', 'SERPINA1', 'F13A1', 'LOC1...</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Platelet Activation, Signaling And Aggregation</td>\n",
       "      <td>R-HSA-76002</td>\n",
       "      <td>1.140647e-25</td>\n",
       "      <td>1.741388e-23</td>\n",
       "      <td>21.719027</td>\n",
       "      <td>1247.389535</td>\n",
       "      <td>['LGALS3BP', 'SERPINA3', 'ITIH3', 'ECM1', 'ORM1', 'SERPINA1', 'F13A1', 'LOC1...</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          pathway    pathway_id          pval           fdr  \\\n",
       "0                          Platelet Degranulation  R-HSA-114608  4.890639e-33  2.239913e-30   \n",
       "1    Response To Elevated Platelet Cytosolic Ca2+   R-HSA-76005  1.555951e-32  3.563128e-30   \n",
       "2  Platelet Activation, Signaling And Aggregation   R-HSA-76002  1.140647e-25  1.741388e-23   \n",
       "\n",
       "   odds_ratio  combined_score  \\\n",
       "0   48.182590     3584.687599   \n",
       "1   45.832030     3356.767042   \n",
       "2   21.719027     1247.389535   \n",
       "\n",
       "                                                                             genes  num_of_genes  \n",
       "0  ['LGALS3BP', 'SERPINA3', 'ECM1', 'ORM1', 'ITIH3', 'SERPINA1', 'F13A1', 'LOC1...            27  \n",
       "1  ['LGALS3BP', 'SERPINA3', 'ECM1', 'ORM1', 'ITIH3', 'SERPINA1', 'F13A1', 'LOC1...            27  \n",
       "2  ['LGALS3BP', 'SERPINA3', 'ITIH3', 'ECM1', 'ORM1', 'SERPINA1', 'F13A1', 'LOC1...            28  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bpx.df_enr.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1f1dc924-e9aa-4668-ad26-46e57a79b12e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "458"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_enr0=bpx.df_enr0\n",
    "len(df_enr0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ce3e688-8b03-47ca-a747-180e47f4cf2f",
   "metadata": {},
   "source": [
    "### Instantiating Gemini"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "97e9c4fe-332c-45db-8e99-a82ac23f804e",
   "metadata": {},
   "outputs": [],
   "source": [
    "##################\n",
    "is_seldata=True\n",
    "i_dfp=0\n",
    "i_dfp_list=[0]\n",
    "##################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8aac5168-2838-45ca-93e3-d5896cff735e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).\n",
      "Using a pipeline without specifying a model name and revision in production is not recommended.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "COVID-19 True [0] [1, 3]\n",
      "Context: COVID-19 is the disease caused by the virus SARS-CoV-2. When the virus enters your body, it infects cells, primarily in the respiratory system.\n"
     ]
    }
   ],
   "source": [
    "API_KEY = dic_yml['API_KEY']\n",
    "\n",
    "disease = dic_yml['disease']\n",
    "context_disease = dic_yml['context_disease']\n",
    "n_sentences = dic_yml['n_sentences']\n",
    "chosen_model_sampling = dic_yml['chosen_model_sampling']\n",
    "\n",
    "gem = Gemini( bpx=bpx, is_seldata=is_seldata, disease=disease, context_disease=context_disease, \n",
    "             API_KEY=API_KEY, n_sentences=n_sentences, root0=root0, \n",
    "             chosen_model_list=chosen_model_list, i_dfp_list=i_dfp_list, chosen_model_sampling=chosen_model_sampling)\n",
    "print(\"\\n\")\n",
    "print(gem.disease, gem.is_seldata, gem.i_dfp_list, gem.chosen_model_list)\n",
    "print(\"Context:\", context_disease)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "66e3cb4b-4611-4a5d-b57c-6dffab8ed55e",
   "metadata": {},
   "outputs": [],
   "source": [
    "gem.set_case(bpx.case, bpx.df_enr, bpx.df_enr0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fcb9b5b-4491-4065-af4a-3859ac1c41c2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6bcf5f04-74b6-4de3-b118-8a64bc423d3a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start opening tables ....\n",
      "Building synonym dictionary ...\n",
      "\n",
      "File ../src/down_pdf_pmid.sh exists True\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "terms1_param = dic_yml['terms1_param']\n",
    "terms2_param = dic_yml['terms2_param']\n",
    "terms_not_param = dic_yml['terms_not_param']\n",
    "connective_param = dic_yml['connective_param']\n",
    "remove_synonym_list = dic_yml['remove_synonym_list']\n",
    "inidate = dic_yml['inidate']\n",
    "enddate = dic_yml['enddate']\n",
    "verbose_query = dic_yml['verbose_query']\n",
    "force_query = dic_yml['force_query']\n",
    "sleep_entrez = dic_yml['sleep_entrez']\n",
    "retmax = dic_yml['retmax']\n",
    "try_all_text = dic_yml['try_all_text']\n",
    "text_quote = dic_yml['text_quote']\n",
    "dec_ncpus = dic_yml['dec_ncpus']\n",
    "sleep_TIKA = dic_yml['sleep_TIKA']\n",
    "min_words_text = dic_yml['min_words_text']\n",
    "\n",
    "prefix = s_project\n",
    "\n",
    "pub = Pubmed(bpx, gem, email, prefix, root0=root0, \n",
    "             inidate=inidate, enddate=enddate, \n",
    "             terms1_param=terms1_param, terms2_param=terms2_param,\n",
    "             terms_not_param=terms_not_param, connective_param=connective_param,\n",
    "             remove_synonym_list=remove_synonym_list, \n",
    "             sleep_entrez = sleep_entrez, retmax=retmax,  \n",
    "             try_all_text=try_all_text, text_quote=text_quote,\n",
    "             root_colab=root_colab, dec_ncpus=dec_ncpus, sleep_TIKA=sleep_TIKA, min_words_text=min_words_text)\n",
    "\n",
    "pub.is_seldata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17880398-d3dd-44fa-b49b-4f96f329e227",
   "metadata": {},
   "source": [
    "### Settings: selected data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3e6701a1-1b75-4e2a-b870-cddeceedc990",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1, 3]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chosen_model_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "526354f3-6c11-461e-b792-be51aac3abcd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Table opened ((30, 3)) at '../../colaboracoes/covid/sonia_andrade/taubate/proteomics_202205/gemini/sampling_30_regs_yes_no_case_g3_male_adult_model_gemini-1.5-flash_query_type_strong.tsv'\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway</th>\n",
       "      <th>fdr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>R-HSA-8953854</td>\n",
       "      <td>Metabolism Of RNA</td>\n",
       "      <td>0.906000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>R-HSA-194068</td>\n",
       "      <td>Bile Acid And Bile Salt Metabolism</td>\n",
       "      <td>0.255000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R-HSA-9029569</td>\n",
       "      <td>NR1H3 And NR1H2 Regulate Gene Expression Linked To Cholesterol Transport And...</td>\n",
       "      <td>0.000128</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      pathway_id                                                                          pathway  \\\n",
       "0  R-HSA-8953854                                                                Metabolism Of RNA   \n",
       "1   R-HSA-194068                                               Bile Acid And Bile Salt Metabolism   \n",
       "2  R-HSA-9029569  NR1H3 And NR1H2 Regulate Gene Expression Linked To Cholesterol Transport And...   \n",
       "\n",
       "        fdr  \n",
       "0  0.906000  \n",
       "1  0.255000  \n",
       "2  0.000128  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test=False\n",
    "force=False\n",
    "verbose=False\n",
    "\n",
    "chosen_model=1\n",
    "gemini_model='gemini-1.5-pro'\n",
    "\n",
    "case_sel_list = ['g3_male_adult', 'g3_female_elder']\n",
    "case_list = case_sel_list\n",
    "case_sel0 = case_sel_list[0]\n",
    "case_sel1 = case_sel_list[1]\n",
    "\n",
    "with_gender=True\n",
    "\n",
    "query_type='strong'\n",
    "N=30\n",
    "\n",
    "############## for selected ###############\n",
    "pub.gem.chosen_model_list = chosen_model_list\n",
    "#############################\n",
    "\n",
    "case=case_sel_list[0]\n",
    "\n",
    "print(\"\")\n",
    "dfsel = pub.gem.open_yes_no_sampling(case=case, N=N, query_type=query_type, verbose=True)\n",
    "print(\"\")\n",
    "dfsel.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "fcd3901a-213f-4595-960c-d524c341c2cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(True, True, ['run01', 'run02'], [1, 3], 3, [0], 5, 3)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pub.gem.is_seldata, pub.is_seldata,run_list, chosen_model_list, chosen_model, i_dfp_list, pub.gem.n_sentences, pub.gem.chosen_model_sampling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33abaae5-1752-4265-9e17-193fdef0bd4e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f970f700-7573-42e1-a78c-7c58e3aaacfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table opened ((2673, 2)) at '../../colaboracoes/reactome/data/ReactomePathways_hsa.tsv'\n",
      "2673\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2670</th>\n",
       "      <td>R-HSA-199992</td>\n",
       "      <td>trans-Golgi Network Vesicle Budding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2671</th>\n",
       "      <td>R-HSA-192814</td>\n",
       "      <td>vRNA Synthesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2672</th>\n",
       "      <td>R-HSA-192905</td>\n",
       "      <td>vRNP Assembly</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        pathway_id                              pathway\n",
       "2670  R-HSA-199992  trans-Golgi Network Vesicle Budding\n",
       "2671  R-HSA-192814                       vRNA Synthesis\n",
       "2672  R-HSA-192905                        vRNP Assembly"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfr=gem.reactome.open_reactome(verbose=True)\n",
    "print(len(dfr))\n",
    "dfr.tail(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "865cd64f-1d4d-4978-bc87-cfa5a475c043",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1780</th>\n",
       "      <td>R-HSA-71406</td>\n",
       "      <td>Pyruvate metabolism and Citric Acid (TCA) cycle</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       pathway_id                                          pathway\n",
       "1780  R-HSA-71406  Pyruvate metabolism and Citric Acid (TCA) cycle"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway_id='R-HSA-71406'\n",
    "df2=dfr[dfr.pathway_id == pathway_id]\n",
    "df2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84bd5b81-9b5a-4be3-9579-add96b2d6361",
   "metadata": {},
   "source": [
    "### Reactome abstracts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "0efb994e-bd68-4676-b92f-1167cc90f3fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'cdc25A protein is degraded by the ubiquitin-proteasome machinery in both terminally differentiating and cycling cells (Bernardi et al. 2000).'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "i=0\n",
    "\n",
    "row=gem.dfr.iloc[i]\n",
    "pathway=row.pathway\n",
    "ptw_abst=row.abstract\n",
    "ptw_abst2=gem.prepare_abstract_n_sentences(ptw_abst)\n",
    "ptw_abst2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82eb985b-a6b8-4c31-b7ee-925825425905",
   "metadata": {},
   "source": [
    "### Starting run all SELECTED PATHWAYS - save in pubmed directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f37b09ca-f1d1-4b1f-9235-62731fb98540",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(True, ['run01', 'run02'], ['g3_male_adult', 'g3_female_elder'], [0], 3)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pub.is_seldata, run_list, case_sel_list, i_dfp_list, chosen_model_sampling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "d397d5ad-41b0-4926-aa1b-ede909fbceec",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>> run01\n",
      "\tg3_male_adult model 1 >> Gemini model: gemini-1.5-pro\n",
      "\t$$$ iq=0 simple \t\tdfp 0 # 30\n",
      "\t$$$ iq=1 simple+pubmed \t\tdfp 0 # 30\n",
      "\t$$$ iq=2 disease \t\tdfp 0 # 30\n",
      "\t$$$ iq=3 disease+pubmed \t\tdfp 0 # 30\n",
      "\n",
      "-------------- final end --------------\n",
      "\n",
      "\n",
      "\tg3_female_elder model 1 >> Gemini model: gemini-1.5-pro\n",
      "\t$$$ iq=0 simple \t\tdfp 0 # 30\n",
      "\t$$$ iq=1 simple+pubmed \t\tdfp 0 # 30\n",
      "\t$$$ iq=2 disease \t\tdfp 0 # 30\n",
      "\t$$$ iq=3 disease+pubmed \t\tdfp 0 # 30\n",
      "\n",
      "-------------- final end --------------\n",
      "\n",
      "\n",
      "\n",
      "====================================\n",
      "\n",
      ">>> run02\n",
      "\tg3_male_adult model 1 >> Gemini model: gemini-1.5-pro\n",
      "\t$$$ iq=0 simple \t\tdfp 0 # 30\n",
      "\t$$$ iq=1 simple+pubmed \t\tdfp 0 # 30\n",
      "\t$$$ iq=2 disease \t\tdfp 0 # 30\n",
      "\t$$$ iq=3 disease+pubmed \t\tdfp 0 # 30\n",
      "\n",
      "-------------- final end --------------\n",
      "\n",
      "\n",
      "\tg3_female_elder model 1 >> Gemini model: gemini-1.5-pro\n",
      "\t$$$ iq=0 simple \t\tdfp 0 # 30\n",
      "\t$$$ iq=1 simple+pubmed \t\tdfp 0 # 30\n",
      "\t$$$ iq=2 disease \t\tdfp 0 # 30\n",
      "\t$$$ iq=3 disease+pubmed \t\tdfp 0 # 30\n",
      "\n",
      "-------------- final end --------------\n",
      "\n",
      "\n",
      "\n",
      "====================================\n",
      "\n",
      "CPU times: user 1.19 s, sys: 84.2 ms, total: 1.27 s\n",
      "Wall time: 1.27 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "force=False\n",
    "verbose=False\n",
    "num_tries=3\n",
    "pause_secs=0\n",
    "\n",
    "chosen_model=1\n",
    "\n",
    "for run in run_list:\n",
    "    print(\">>>\", run,)\n",
    "    for case in case_sel_list:\n",
    "        print(f'\\t{case} model {chosen_model}', end=' ')\n",
    "        gem.run_all_selected_gemini(run=run, case=case, i_dfp_list=i_dfp_list, chosen_model=chosen_model, \n",
    "                                    N=N, query_type=query_type,\n",
    "                                    num_tries=num_tries, pause_secs=pause_secs,\n",
    "                                    force=force, verbose=verbose)\n",
    "        print(\"\")\n",
    "    print(\"\\n====================================\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2186af1e-76c4-4d69-a441-3bdbc9892dc6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "34d163c2-de6d-4386-919e-35d561d7f44a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table opened ((30, 11)) at '../../colaboracoes/covid/sonia_andrade/taubate/proteomics_202205/pubgem/gemini/run01/gemini_search_for_COVID-19_g3_male_adult_question_with_PubMed_yes_no_possible_low_evidence_question_strong_relationship_in_studies_of_modulated_pathways_0_selected_model_gemini-1.5-pro_selected.tsv'\n",
      "30\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run</th>\n",
       "      <th>case</th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway</th>\n",
       "      <th>i_dfp</th>\n",
       "      <th>fdr</th>\n",
       "      <th>curation</th>\n",
       "      <th>response_explain</th>\n",
       "      <th>score_explain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>run01</td>\n",
       "      <td>g3_male_adult</td>\n",
       "      <td>R-HSA-114608</td>\n",
       "      <td>Platelet Degranulation</td>\n",
       "      <td>0</td>\n",
       "      <td>6.980000e-10</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes; Several studies in PubMed demonstrate a strong relationship between pla...</td>\n",
       "      <td>0.989471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>run01</td>\n",
       "      <td>g3_male_adult</td>\n",
       "      <td>R-HSA-8964058</td>\n",
       "      <td>HDL Remodeling</td>\n",
       "      <td>0</td>\n",
       "      <td>8.930000e-07</td>\n",
       "      <td>Low</td>\n",
       "      <td>Low evidence.\\n\\nWhile HDL is known to play a role in lipid metabolism and i...</td>\n",
       "      <td>-0.990575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>run01</td>\n",
       "      <td>g3_male_adult</td>\n",
       "      <td>R-HSA-6798695</td>\n",
       "      <td>Neutrophil Degranulation</td>\n",
       "      <td>0</td>\n",
       "      <td>9.760000e-06</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes; Neutrophil degranulation is strongly implicated in severe COVID-19, par...</td>\n",
       "      <td>0.903832</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     run           case     pathway_id                   pathway  i_dfp           fdr curation  \\\n",
       "0  run01  g3_male_adult   R-HSA-114608    Platelet Degranulation      0  6.980000e-10      Yes   \n",
       "1  run01  g3_male_adult  R-HSA-8964058            HDL Remodeling      0  8.930000e-07      Low   \n",
       "2  run01  g3_male_adult  R-HSA-6798695  Neutrophil Degranulation      0  9.760000e-06      Yes   \n",
       "\n",
       "                                                                  response_explain  score_explain  \n",
       "0  Yes; Several studies in PubMed demonstrate a strong relationship between pla...       0.989471  \n",
       "1  Low evidence.\\n\\nWhile HDL is known to play a role in lipid metabolism and i...      -0.990575  \n",
       "2  Yes; Neutrophil degranulation is strongly implicated in severe COVID-19, par...       0.903832  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "verbose=True\n",
    "\n",
    "case=case_sel_list[0]\n",
    "run='run01'\n",
    "query_type='_strong'\n",
    "want_pubmed=True\n",
    "\n",
    "dfgem = gem.get_gemini_results_by_case_model_semantics(run=run, case=case, chosen_model=chosen_model,\n",
    "                                                       i_dfp_list=i_dfp_list, want_pubmed=want_pubmed, \n",
    "                                                       query_type=query_type, verbose=verbose)\n",
    "if dfgem is None:\n",
    "    dfgem = pd.DataFrame()\n",
    "    \n",
    "print(len(dfgem))\n",
    "dfgem.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c60417f-4a52-4b5c-8160-78a060557f1e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
