{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "dd0f894e-f206-4051-96dd-d196f1954d1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.11.11\n"
     ]
    }
   ],
   "source": [
    "from platform import python_version\n",
    "print(python_version())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "693974f9-d0e9-43dd-9256-5b70c1a19a2e",
   "metadata": {},
   "source": [
    "## Gemini API\n",
    "\n",
    "https://ai.google.dev/gemini-api/docs\n",
    "\n",
    "#### API key - Free of charge\n",
    "\n",
    "https://aistudio.google.com/app/apikey\n",
    "\n",
    "AIzaSyA1ZXcSe6NP5jiIw93sUpZYb8RKK1PgYDE\n",
    "\n",
    "### Google Enable API\n",
    "\n",
    "  - You are about to enable 'Generative Language API'.\n",
    "\n",
    "https://ai.google.dev/gemini-api/docs/oauth"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba93597d-d3ad-4e27-a563-6a8276cb35ae",
   "metadata": {},
   "source": [
    "### Costs & Billing\n",
    "\n",
    "https://console.cloud.google.com/billing/01C02C-666E6E-D731B9?project=gen-lang-client-0516343733\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd0fc3a5-3f5a-4bd9-8d67-84ef0275adf3",
   "metadata": {},
   "source": [
    "### Google Python projects\n",
    "\n",
    "#### Gemini API Text Implementation\n",
    "\n",
    "https://github.com/RepellentSpy/Gemini-API-Text-Implementation/tree/main\n",
    "\n",
    "#### gemini-api 0.1.6\n",
    "\n",
    "https://pypi.org/project/gemini-api/\n",
    "\n",
    "\n",
    "#### Gemini-API\n",
    "\n",
    "https://github.com/dsdanielpark/Gemini-API"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2069d3bb-c1e2-4598-9156-a517ad2e1c2b",
   "metadata": {},
   "source": [
    "## LLM - Large Language Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "41a4f889-2fc7-4ea1-90ab-e1efbf15ad0d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import torch\n",
    "torch.cuda.is_available()\n",
    "# !nvidia-smi\n",
    "# !lsmod | grep nvidia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "24db0cf4-e681-493b-9208-f832bdf0c175",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>:root { --jp-notebook-max-width: 100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import os, sys, pickle\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.set_option('display.width', 100)\n",
    "pd.set_option('max_colwidth', 80)\n",
    "pd.set_option(\"display.precision\", 3)\n",
    "\n",
    "import yaml\n",
    "\n",
    "import seaborn as sns\n",
    "sns.set_context(\"notebook\", font_scale=1.4)\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "%matplotlib inline\n",
    "\n",
    "sys.path.insert(1, '../src/')\n",
    "\n",
    "from Basic import *\n",
    "from biopax_lib import *\n",
    "from gemini_lib import *\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "from IPython.display import display, HTML\n",
    "display(HTML(\"<style>:root { --jp-notebook-max-width: 100% !important; }</style>\"))\n",
    "\n",
    "# !pip3 install pyyaml\n",
    "with open('params.yml', 'r') as file:\n",
    "    dic_yml=yaml.safe_load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "db0e17e2-3775-4709-9873-2f918e1ed275",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "G/P LFC cutoffs: lfc=1.000; fdr=0.050\n",
      "Pathway cutoffs: pval=0.050; fdr=0.050; num of genes=3\n"
     ]
    }
   ],
   "source": [
    "root0=dic_yml['root0']\n",
    "email=dic_yml['email']\n",
    "\n",
    "project=dic_yml['project']\n",
    "s_project=dic_yml['s_project']\n",
    "\n",
    "gene_protein=dic_yml['gene_protein']\n",
    "s_omics=dic_yml['s_omics']\n",
    "\n",
    "has_age=dic_yml['has_age']\n",
    "has_gender=dic_yml['has_gender']\n",
    "\n",
    "want_normalized=dic_yml['want_normalized']\n",
    "\n",
    "abs_lfc_cutoff_inf=dic_yml['abs_lfc_cutoff_inf']\n",
    "s_pathw_enrichm_method=dic_yml['s_pathw_enrichm_method']\n",
    "num_min_degs_for_ptw_enr=dic_yml['num_min_degs_for_ptw_enr']\n",
    "\n",
    "tolerance_pathway_index=dic_yml['tolerance_pathway_index']\n",
    "type_sat_ptw_index=dic_yml['type_sat_ptw_index']\n",
    "saturation_lfc_index=dic_yml['saturation_lfc_index']\n",
    "chosen_model_sampling=dic_yml['chosen_model_sampling']\n",
    "\n",
    "case_list=dic_yml['case_list']\n",
    "\n",
    "pval_pathway_cutoff=dic_yml['pval_pathway_cutoff']\n",
    "fdr_pathway_cutoff=dic_yml['fdr_pathway_cutoff']\n",
    "num_of_genes_cutoff=dic_yml['num_of_genes_cutoff']\n",
    "\n",
    "run_list=dic_yml['run_list']\n",
    "chosen_model_list=dic_yml['chosen_model_list']\n",
    "i_dfp_list=dic_yml['i_dfp_list']\n",
    "\n",
    "exp_normalization='quantile_norm' if want_normalized else None\n",
    "normalization='not_normalized' if exp_normalization is None else exp_normalization\n",
    "\n",
    "cfg=Config(project, s_project, case_list, root0)\n",
    "\n",
    "case=case_list[0]\n",
    "\n",
    "n_genes_annot_ptw, n_degs, n_degs_in_ptw, n_degs_not_in_ptw, degs_in_all_ratio=-1,-1,-1,-1,-1\n",
    "abs_lfc_cutoff, fdr_lfc_cutoff, n_degs, n_degs_up, n_degs_dw=cfg.get_best_lfc_cutoff(case, 'not_normalized')\n",
    "\n",
    "\n",
    "print(f\"G/P LFC cutoffs: lfc={abs_lfc_cutoff:.3f}; fdr={fdr_lfc_cutoff:.3f}\")\n",
    "print(f\"Pathway cutoffs: pval={pval_pathway_cutoff:.3f}; fdr={fdr_pathway_cutoff:.3f}; num of genes={num_of_genes_cutoff}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "75d9a3d5-6a54-4b70-9dd2-deb3ddb09810",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start opening tables ....\n",
      "Building synonym dictionary ...\n",
      "\n",
      "\n",
      "Echo Parameters:\n",
      "For case 'g2a_male', there are 42/42 DAPs/DAPs with ensembl_id\n",
      "DAP's cutoffs: abs(LFC)=0.900; FDR=0.150\n",
      "\t42/42 DAPs/ensembl.\n",
      "\t\tUp 26/26 DAPs/ensembl.\n",
      "\t\tDw 16/16 DAPs/ensembl.\n",
      "\n",
      "Found 45 (best=45) pathways for geneset num=0 'Reactome_2022'\n",
      "Pathway cutoffs p-value=0.050 fdr=0.050 min genes=3\n",
      "DAPs found in enriched pathways:\n",
      "\tThere are 42 DAPs found in pathways\n",
      "\t27 (best=27) DAPs in pathways and 15/15 DAPs/ensembl not in pathways\n",
      "\n",
      "\t20 DAPs ensembl Up in pathways\n",
      "\t6 DAPs Up ensembl not in pathways\n",
      "\n",
      "\t7 DAPs ensembl Dw in pathways\n",
      "\t9 DAPs Dw ensembl not in pathways\n"
     ]
    }
   ],
   "source": [
    "bpx=Biopax(gene_protein, s_omics, project, s_project, root0,\n",
    "           case_list, has_age, has_gender, clone_objects=False,\n",
    "           exp_normalization=exp_normalization, geneset_num=0, \n",
    "           num_min_degs_for_ptw_enr=num_min_degs_for_ptw_enr, \n",
    "           tolerance_pathway_index=tolerance_pathway_index, \n",
    "           s_pathw_enrichm_method=s_pathw_enrichm_method,\n",
    "           abs_lfc_cutoff_inf=abs_lfc_cutoff_inf, \n",
    "           type_sat_ptw_index=type_sat_ptw_index, saturation_lfc_index=saturation_lfc_index)\n",
    "\n",
    "case=case_list[0]\n",
    "\n",
    "bpx.cfg.set_default_best_lfc_cutoff(normalization, abs_lfc_cutoff=1, fdr_lfc_cutoff=0.05)\n",
    "ret, degs, degs_ensembl, dfdegs=bpx.open_case(case, verbose=False)\n",
    "print(\"\\nEcho Parameters:\")\n",
    "bpx.echo_parameters()\n",
    "\n",
    "geneset_num=bpx.geneset_num"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "dacf3997-f558-4b3d-878b-4538b05b5d61",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('enricher_Reactome_2022_taubate_covid19_proteomics_for_g2a_male_x_ctrl_not_normalized_cutoff_lfc_0.900_fdr_0.150.tsv',\n",
       " 'enricher_Reactome_2022_taubate_covid19_proteomics_for_g2a_male_x_ctrl_not_normalized_cutoff_lfc_0.900_fdr_0.150_pathway_pval_0.050_fdr_0.050_num_genes_3.tsv')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fname, fname_cutoff=bpx.set_enrichment_name()\n",
    "fname, fname_cutoff"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e5ef1c6-da9c-40c5-8e8c-651600c3b974",
   "metadata": {},
   "source": [
    "### Run all, after finding a case\n",
    "  - please open dfr (reactome) before"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "39e06289-0bce-414c-aa60-a707526fc8e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "45"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Nenr=len(bpx.df_enr)\n",
    "Nenr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "cb93f0de-13aa-4a71-804c-5f198f9d7997",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway</th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pval</th>\n",
       "      <th>fdr</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>genes</th>\n",
       "      <th>num_of_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Platelet Degranulation</td>\n",
       "      <td>R-HSA-114608</td>\n",
       "      <td>1.952e-17</td>\n",
       "      <td>3.425e-15</td>\n",
       "      <td>70.248</td>\n",
       "      <td>2702.787</td>\n",
       "      <td>['FGB', 'FGA', 'ORM1', 'CLEC3B', 'VWF', 'AHSG', 'FGG', 'SERPING1', 'FLNA', '...</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Response To Elevated Platelet Cytosolic Ca2+</td>\n",
       "      <td>R-HSA-76005</td>\n",
       "      <td>3.172e-17</td>\n",
       "      <td>3.425e-15</td>\n",
       "      <td>67.254</td>\n",
       "      <td>2554.970</td>\n",
       "      <td>['FGB', 'FGA', 'ORM1', 'CLEC3B', 'VWF', 'AHSG', 'FGG', 'SERPING1', 'FLNA', '...</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Platelet Activation, Signaling And Aggregation</td>\n",
       "      <td>R-HSA-76002</td>\n",
       "      <td>7.631e-17</td>\n",
       "      <td>5.494e-15</td>\n",
       "      <td>41.079</td>\n",
       "      <td>1524.518</td>\n",
       "      <td>['FGB', 'FGA', 'ORM1', 'VWF', 'AHSG', 'FGG', 'PPBP', 'GP5', 'YWHAZ', 'CLEC3B...</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          pathway    pathway_id       pval        fdr  odds_ratio  \\\n",
       "0                          Platelet Degranulation  R-HSA-114608  1.952e-17  3.425e-15      70.248   \n",
       "1    Response To Elevated Platelet Cytosolic Ca2+   R-HSA-76005  3.172e-17  3.425e-15      67.254   \n",
       "2  Platelet Activation, Signaling And Aggregation   R-HSA-76002  7.631e-17  5.494e-15      41.079   \n",
       "\n",
       "   combined_score  \\\n",
       "0        2702.787   \n",
       "1        2554.970   \n",
       "2        1524.518   \n",
       "\n",
       "                                                                             genes  num_of_genes  \n",
       "0  ['FGB', 'FGA', 'ORM1', 'CLEC3B', 'VWF', 'AHSG', 'FGG', 'SERPING1', 'FLNA', '...            12  \n",
       "1  ['FGB', 'FGA', 'ORM1', 'CLEC3B', 'VWF', 'AHSG', 'FGG', 'SERPING1', 'FLNA', '...            12  \n",
       "2  ['FGB', 'FGA', 'ORM1', 'VWF', 'AHSG', 'FGG', 'PPBP', 'GP5', 'YWHAZ', 'CLEC3B...            14  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bpx.df_enr.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1f1dc924-e9aa-4668-ad26-46e57a79b12e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "216"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_enr0=bpx.df_enr0\n",
    "len(df_enr0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "523dd881-9d80-4dd9-8488-d87cb354b0a3",
   "metadata": {},
   "source": [
    "### Ensemble: is_seldata=False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e4389cbe-b7c1-433f-a75b-83d5f02acab9",
   "metadata": {},
   "outputs": [],
   "source": [
    "is_seldata=False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8aac5168-2838-45ca-93e3-d5896cff735e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).\n",
      "Using a pipeline without specifying a model name and revision in production is not recommended.\n",
      "Device set to use cuda:0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "COVID-19 False\n",
      "Context: COVID-19 is the disease caused by the virus SARS-CoV-2. When the virus enters your body, it infects cells, primarily in the respiratory system.\n"
     ]
    }
   ],
   "source": [
    "API_KEY=dic_yml['API_KEY']\n",
    "\n",
    "disease=dic_yml['disease']\n",
    "context_disease=dic_yml['context_disease']\n",
    "n_sentences=dic_yml['n_sentences']\n",
    "chosen_model_sampling=dic_yml['chosen_model_sampling']\n",
    "\n",
    "gem=Gemini(bpx=bpx, is_seldata=is_seldata, disease=disease, context_disease=context_disease, \n",
    "           API_KEY=API_KEY, n_sentences=n_sentences, root0=root0, \n",
    "           chosen_model_list=chosen_model_list, i_dfp_list=i_dfp_list, chosen_model_sampling=chosen_model_sampling)\n",
    "print(\"\\n\")\n",
    "print(gem.disease, gem.is_seldata)\n",
    "print(\"Context:\", context_disease)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "186f7ae3-619a-429a-937a-3fdf2c299566",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'<my_gemini_key>'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "API_KEY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "66e3cb4b-4611-4a5d-b57c-6dffab8ed55e",
   "metadata": {},
   "outputs": [],
   "source": [
    "gem.set_case(bpx.case, bpx.df_enr, bpx.df_enr0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "187a7190-2055-4456-8563-5259fdf57af3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f970f700-7573-42e1-a78c-7c58e3aaacfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table opened ((2673, 32)) at '../../../data_aux/reactome/reactome_pathways_abstract.tsv'\n",
      "2673\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbId</th>\n",
       "      <th>pathway</th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>stIdVersion</th>\n",
       "      <th>isInDisease</th>\n",
       "      <th>isInferred</th>\n",
       "      <th>name</th>\n",
       "      <th>releaseDate</th>\n",
       "      <th>speciesName</th>\n",
       "      <th>literatureReference</th>\n",
       "      <th>...</th>\n",
       "      <th>doi</th>\n",
       "      <th>compartment</th>\n",
       "      <th>goBiologicalProcess</th>\n",
       "      <th>normalPathway</th>\n",
       "      <th>previousReviewStatus</th>\n",
       "      <th>crossReference</th>\n",
       "      <th>releaseStatus</th>\n",
       "      <th>definition</th>\n",
       "      <th>precedingEvent</th>\n",
       "      <th>relatedSpecies</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2670</th>\n",
       "      <td>192905.0</td>\n",
       "      <td>vRNP Assembly</td>\n",
       "      <td>R-HSA-192905</td>\n",
       "      <td>R-HSA-192905.5</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>['vRNP Assembly']</td>\n",
       "      <td>2007-02-28</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[{'dbId': 192782, 'displayName': 'Nuclear traffic of influenza virus protein...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{'dbId': 7660, 'displayName': 'nucleoplasm', 'accession': '0005654', 'datab...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{'dbId': 9031716, 'displayName': 'Influenza A virus', 'name': ['Influenza A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2671</th>\n",
       "      <td>139910.0</td>\n",
       "      <td>Activation of BMF and translocation to mitochondria</td>\n",
       "      <td>R-HSA-139910</td>\n",
       "      <td>R-HSA-139910.5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>['Activation of BMF and translocation to mitochondria']</td>\n",
       "      <td>2004-10-27</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[{'dbId': 140368, 'displayName': 'The Bcl-2 family: roles in cell survival a...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{'dbId': 70101, 'displayName': 'cytosol', 'accession': '0005829', 'database...</td>\n",
       "      <td>{'dbId': 5673453, 'displayName': 'positive regulation of establishment of pr...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2672</th>\n",
       "      <td>532668.0</td>\n",
       "      <td>N-glycan trimming in the ER and Calnexin/Calreticulin cycle</td>\n",
       "      <td>R-HSA-532668</td>\n",
       "      <td>R-HSA-532668.4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>['N-glycan trimming in the ER and Calnexin/Calreticulin cycle']</td>\n",
       "      <td>2010-09-21</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[{'dbId': 901083, 'displayName': 'Getting in and out from calnexin/calreticu...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'dbId': 7977, 'displayName': 'protein folding', 'accession': '0006457', 'da...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          dbId                                                      pathway    pathway_id  \\\n",
       "2670  192905.0                                                vRNP Assembly  R-HSA-192905   \n",
       "2671  139910.0          Activation of BMF and translocation to mitochondria  R-HSA-139910   \n",
       "2672  532668.0  N-glycan trimming in the ER and Calnexin/Calreticulin cycle  R-HSA-532668   \n",
       "\n",
       "         stIdVersion isInDisease isInferred  \\\n",
       "2670  R-HSA-192905.5        True      False   \n",
       "2671  R-HSA-139910.5       False      False   \n",
       "2672  R-HSA-532668.4       False      False   \n",
       "\n",
       "                                                                 name releaseDate   speciesName  \\\n",
       "2670                                                ['vRNP Assembly']  2007-02-28  Homo sapiens   \n",
       "2671          ['Activation of BMF and translocation to mitochondria']  2004-10-27  Homo sapiens   \n",
       "2672  ['N-glycan trimming in the ER and Calnexin/Calreticulin cycle']  2010-09-21  Homo sapiens   \n",
       "\n",
       "                                                                  literatureReference  ...  doi  \\\n",
       "2670  [{'dbId': 192782, 'displayName': 'Nuclear traffic of influenza virus protein...  ...  NaN   \n",
       "2671  [{'dbId': 140368, 'displayName': 'The Bcl-2 family: roles in cell survival a...  ...  NaN   \n",
       "2672  [{'dbId': 901083, 'displayName': 'Getting in and out from calnexin/calreticu...  ...  NaN   \n",
       "\n",
       "                                                                          compartment  \\\n",
       "2670  [{'dbId': 7660, 'displayName': 'nucleoplasm', 'accession': '0005654', 'datab...   \n",
       "2671  [{'dbId': 70101, 'displayName': 'cytosol', 'accession': '0005829', 'database...   \n",
       "2672                                                                              NaN   \n",
       "\n",
       "                                                                  goBiologicalProcess  \\\n",
       "2670                                                                              NaN   \n",
       "2671  {'dbId': 5673453, 'displayName': 'positive regulation of establishment of pr...   \n",
       "2672  {'dbId': 7977, 'displayName': 'protein folding', 'accession': '0006457', 'da...   \n",
       "\n",
       "     normalPathway previousReviewStatus crossReference releaseStatus definition precedingEvent  \\\n",
       "2670           NaN                  NaN            NaN           NaN        NaN            NaN   \n",
       "2671           NaN                  NaN            NaN           NaN        NaN            NaN   \n",
       "2672           NaN                  NaN            NaN           NaN        NaN            NaN   \n",
       "\n",
       "                                                                       relatedSpecies  \n",
       "2670  [{'dbId': 9031716, 'displayName': 'Influenza A virus', 'name': ['Influenza A...  \n",
       "2671                                                                              NaN  \n",
       "2672                                                                              NaN  \n",
       "\n",
       "[3 rows x 32 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfr=gem.reactome.open_reactome_abstract(verbose=True)\n",
    "print(len(dfr))\n",
    "dfr.tail(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "865cd64f-1d4d-4978-bc87-cfa5a475c043",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbId</th>\n",
       "      <th>pathway</th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>stIdVersion</th>\n",
       "      <th>isInDisease</th>\n",
       "      <th>isInferred</th>\n",
       "      <th>name</th>\n",
       "      <th>releaseDate</th>\n",
       "      <th>speciesName</th>\n",
       "      <th>literatureReference</th>\n",
       "      <th>...</th>\n",
       "      <th>doi</th>\n",
       "      <th>compartment</th>\n",
       "      <th>goBiologicalProcess</th>\n",
       "      <th>normalPathway</th>\n",
       "      <th>previousReviewStatus</th>\n",
       "      <th>crossReference</th>\n",
       "      <th>releaseStatus</th>\n",
       "      <th>definition</th>\n",
       "      <th>precedingEvent</th>\n",
       "      <th>relatedSpecies</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2625</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Pyruvate metabolism and Citric Acid (TCA) cycle</td>\n",
       "      <td>R-HSA-71406</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      dbId                                          pathway   pathway_id stIdVersion isInDisease  \\\n",
       "2625   NaN  Pyruvate metabolism and Citric Acid (TCA) cycle  R-HSA-71406         NaN         NaN   \n",
       "\n",
       "     isInferred name releaseDate speciesName literatureReference  ...  doi compartment  \\\n",
       "2625        NaN  NaN         NaN         NaN                 NaN  ...  NaN         NaN   \n",
       "\n",
       "     goBiologicalProcess normalPathway previousReviewStatus crossReference releaseStatus  \\\n",
       "2625                 NaN           NaN                  NaN            NaN           NaN   \n",
       "\n",
       "     definition precedingEvent relatedSpecies  \n",
       "2625        NaN            NaN            NaN  \n",
       "\n",
       "[1 rows x 32 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway_id='R-HSA-71406'\n",
    "df2=dfr[dfr.pathway_id == pathway_id]\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "0efb994e-bd68-4676-b92f-1167cc90f3fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'cdc25A protein is degraded by the ubiquitin-proteasome machinery in both terminally differentiating and cycling cells (Bernardi et al. 2000).'"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "i=0\n",
    "\n",
    "row=dfr.iloc[i]\n",
    "pathway=row.pathway\n",
    "ptw_abst=row.abstract\n",
    "ptw_abst2=gem.prepare_abstract_n_sentences(ptw_abst)\n",
    "ptw_abst2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c2bd355-0e2b-495d-8868-7b1cb5afc9b2",
   "metadata": {},
   "source": [
    "### Question type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "6a003d91-57ac-48d1-b32f-1248a20bd759",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "simple\n",
      "without_PubMed and sufix 'yes_no_possible_low_evidence_studies_of_modulated_pathways'\n",
      "is the pathway 'Ubiquitin Mediated Degradation of Phosphorylated Cdc25A' studied about the disease COVID-19 for male adult mild outpatient? \n",
      "\n",
      "\n",
      "simple+pubmed\n",
      "with_PubMed and sufix 'yes_no_possible_low_evidence_studies_of_modulated_pathways'\n",
      "is the pathway 'Ubiquitin Mediated Degradation of Phosphorylated Cdc25A' studied in PubMed, about the disease COVID-19 for male adult mild outpatient? \n",
      "\n",
      "\n",
      "disease\n",
      "without_PubMed and sufix 'yes_no_possible_low_evidence_question_strong_relationship_in_studies_of_modulated_pathways'\n",
      "has the pathway 'Ubiquitin Mediated Degradation of Phosphorylated Cdc25A' a strong relationship in studies related to the disease COVID-19 for male adult mild outpatient? \n",
      "\n",
      "\n",
      "disease+pumed\n",
      "with_PubMed and sufix 'yes_no_possible_low_evidence_question_strong_relationship_in_studies_of_modulated_pathways'\n",
      "has the pathway 'Ubiquitin Mediated Degradation of Phosphorylated Cdc25A' a strong relationship in studies found in PubMed, related to the disease COVID-19 for male adult mild outpatient? \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for quest_type in ['simple', 'simple+pubmed', 'disease', 'disease+pumed'] :\n",
    "    question, with_without_PubMed, sufix=gem.define_question(quest_type)\n",
    "    print(quest_type)\n",
    "    print(f\"{with_without_PubMed} and sufix '{sufix}'\")\n",
    "    print(question%(pathway), '\\n\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6efadbab-d2f8-44a0-a901-55c9c171e037",
   "metadata": {},
   "source": [
    "### gemini model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "b6663755-258a-4b3b-85f5-b143bd6bb3e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['gemini-1.0-pro',\n",
       " 'gemini-1.5-pro',\n",
       " 'gemini-1.5-flash-8b',\n",
       " 'gemini-1.5-flash',\n",
       " 'gemini-2.0-flash-exp',\n",
       " 'gemma-2-2b-it',\n",
       " 'gemma-2-9b-it',\n",
       " 'gemma-2-27b-it']"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gem.gemini_models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "511dfe67-db13-4606-aae6-0c4d81083aa3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'gemini-1.0-pro'"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chosen_model=0\n",
    "gem.set_gemini_num_model(chosen_model)\n",
    "gem.gemini_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "513b5887-d867-4fc9-b019-c0767b52f18d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gemini-1.0-pro\n",
      ".>>> curl_msg\n",
      "curl -H 'Content-Type: application/json' -d '{\"contents\":[{\"parts\":[{\"text\":\"{Answer in the first line Yes, Possible, Low evidence, or No; and explain; is the pathway Pyruvate metabolism and Citric Acid (TCA) cycle studied about the disease COVID-19 for male adult mild outpatient? Context: In the citric acid or tricarboxylic acid (TCA) cycle, the acetyl group of acetyl CoA (derived primarily from oxidative decarboxylation of pyruvate, beta-oxidation of long-chain fatty acids, and catabolism of ketone bodies and several amino acids) can be completely oxidized to CO2 in reactions that also yield one high-energy phosphate bond (as GTP or ATP) and four reducing equivalents (three NADH + H+, and one FADH2). Then, the electron transport chain oxidizes NADH and FADH2 to yield nine more high-energy phosphate bonds (as ATP). All reactions of the citric acid cycle take place in the mitochondrion. And COVID-19 is the disease caused by the virus SARS-CoV-2. When the virus enters your body, it infects cells, primarily in the respiratory system.}\"}]}], \"generationConfig\":{\"temperature\":0.2, \"topK\":50,\"topP\":0.2, \"maxOutputTokens\":4096,\"stopSequences\":[]}}' -X POST '{https://generativelanguage.googleapis.com/v1beta/models/gemini-1.0-pro-latest:generateContent?key=<my_gemini_key>}' --silent >> ./gemini_0/response.json\n",
      "\n",
      "Error: calling gemini '<class 'ValueError'>'\n",
      "."
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "want=True\n",
    "i_try=0\n",
    "\n",
    "print(gem.gemini_model)\n",
    "\n",
    "list_candidates=[]\n",
    "\n",
    "quest_type='simple'\n",
    "print_ok=True\n",
    "\n",
    "if want:\n",
    "    i=0\n",
    "    while True:\n",
    "        print(\".\", end='')\n",
    "\n",
    "        question0, with_without_PubMed, sufix=gem.define_question(quest_type)\n",
    "\n",
    "        question_name0=f'{with_without_PubMed}_{sufix}_0_default'\n",
    "        question_name1=f'{with_without_PubMed}_{sufix}_0_first'\n",
    "        question_name2=f'{with_without_PubMed}_{sufix}_1_middle' \n",
    "        question_name3=f'{with_without_PubMed}_{sufix}_2_final'\n",
    "        question_name4=f'{with_without_PubMed}_{sufix}_3_others'\n",
    "\n",
    "        multiple_data =[ [0, question_name0, gem.df_enr], ]\n",
    "\n",
    "        dfr=gem.dfr[gem.dfr.pathway_id == pathway_id]\n",
    "        if dfr.empty:\n",
    "            print(f\"\\nError: pathway_id {pathway_id} not found in reactome dfr.\")\n",
    "        else:\n",
    "            pathway =dfr.iloc[0].pathway\n",
    "            ptw_abst=dfr.iloc[0].abstract\n",
    "            ptw_abst=gem.prepare_abstract_n_sentences(ptw_abst)\n",
    "    \n",
    "            if ptw_abst[-1] == '.':\n",
    "                ptw_abst=ptw_abst[:-1]\n",
    "    \n",
    "            for i_dfp, question_name, dfp in multiple_data:\n",
    "        \n",
    "                s_question0=question0%(pathway)\n",
    "                question=gem.prefix_question + s_question0 + f\" Context: {ptw_abst}. And {gem.context_disease}\"\n",
    "    \n",
    "                break\n",
    "    \n",
    "            list_candidates=gem.run_curl_gemini(question, temperature=.2, topP=.2, verbose=print_ok)\n",
    "            print_ok=False\n",
    "            \n",
    "            time.sleep(3)\n",
    "            if len(list_candidates) != 0:\n",
    "                break\n",
    "    \n",
    "        i += 1\n",
    "        if i == 2:\n",
    "            break\n",
    "\n",
    "len(list_candidates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "d0346d9a-20c1-444f-9f42-bf67b36485bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if want:\n",
    "    if len(list_candidates) != 0:\n",
    "        response=gem.response_candidate(list_candidates, 0)\n",
    "    else:\n",
    "        response=''\n",
    "else:\n",
    "    response=''\n",
    "\n",
    "Markdown(response)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82eb985b-a6b8-4c31-b7ee-925825425905",
   "metadata": {},
   "source": [
    "### Starting run all\n",
    "\n",
    "#### Runs\n",
    "\n",
    "  - 0 - default cutoff\n",
    "  - 1 - BCA cutoff\n",
    "  - 2 - middle of the table\n",
    "  - 3 - end of the table\n",
    "  - 4 - others random genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "d397d5ad-41b0-4926-aa1b-ede909fbceec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gemini-1.0-pro \n",
      "\n",
      "\n",
      ">> Gemini model: gemini-1.0-pro\n",
      "Please, configure API_KEY in params.yml\n",
      "\n",
      "====================================\n",
      "\n",
      ">> Gemini model: gemini-1.0-pro\n",
      "Please, configure API_KEY in params.yml\n",
      "\n",
      "====================================\n",
      "\n",
      "CPU times: user 477 ms, sys: 20 ms, total: 497 ms\n",
      "Wall time: 495 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "force = False\n",
    "verbose = False\n",
    "num_tries = 3\n",
    "pause_secs = 0\n",
    "\n",
    "print(gem.gemini_model, '\\n\\n')\n",
    "\n",
    "for run in run_list:\n",
    "    gem.run_all_gemini(run=run, case_list=case_list, chosen_model=chosen_model, \n",
    "                       num_tries=num_tries, pause_secs=pause_secs, force=force, verbose=verbose)\n",
    "    print(\"\\n====================================\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76f76921-3e5d-418d-b7b2-97b552e12214",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e11d40d2-3bd2-4dbb-8439-196d1e026a58",
   "metadata": {},
   "outputs": [],
   "source": [
    "verbose=True\n",
    "\n",
    "case=case_list[3]\n",
    "run='run01'\n",
    "chosen_model=3\n",
    "query_type='_strong'\n",
    "want_pubmed=True\n",
    "\n",
    "dfgem=gem.get_gemini_results_by_case_model_semantics(run=run, case=case, chosen_model=chosen_model,\n",
    "                                                     i_dfp_list=i_dfp_list, want_pubmed=want_pubmed, \n",
    "                                                     query_type=query_type, verbose=verbose)\n",
    "if dfgem is None:\n",
    "    dfgem=pd.DataFrame()\n",
    "    \n",
    "print(len(dfgem))\n",
    "dfgem.head(3)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
