{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.12.0\n"
     ]
    }
   ],
   "source": [
    "from platform import python_version\n",
    "print(python_version())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import os, sys, pickle\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.set_option('display.width', 100)\n",
    "pd.set_option('max_colwidth', 80)\n",
    "\n",
    "import seaborn as sns\n",
    "sns.set_context(\"notebook\", font_scale=1.4)\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "%matplotlib inline\n",
    "\n",
    "sys.path.insert(1, '../src/')\n",
    "\n",
    "from Basic import *\n",
    "from entrez_conversion import *\n",
    "from pubmed_lib import *\n",
    "from gemini_lib import *\n",
    "from biopax_lib import *\n",
    "\n",
    "from IPython.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:100% !important; }</style>\"))\n",
    "\n",
    "email = \"flalix@gmail.com\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pubmed search\n",
    "\n",
    "#### scipy corpus data\n",
    "\n",
    "https://allenai.github.io/scispacy/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "G/P LFC cutoffs: lfc=1.000; fdr=0.050\n",
      "Pathway cutoffs: pval=0.050; fdr=0.050; num of genes=3\n"
     ]
    }
   ],
   "source": [
    "root_chibe = \"../../chibe/\"\n",
    "root_colab = '../../colaboracoes/'\n",
    "root0       = '../../colaboracoes/covid/sonia_andrade/taubate/proteomics_202205/'\n",
    "\n",
    "project = 'Taubate COVID-19'\n",
    "s_project = 'taubate_covid19'\n",
    "\n",
    "gene_protein = 'protein'\n",
    "s_omics = 'proteomics'\n",
    "\n",
    "has_age = True\n",
    "has_gender = True\n",
    "\n",
    "want_normalized = False\n",
    "exp_normalization='quantile_norm' if want_normalized else None\n",
    "normalization = 'not_normalized' if exp_normalization is None else exp_normalization\n",
    "\n",
    "abs_lfc_cutoff_inf = 0.40\n",
    "s_pathw_enrichm_method = 'enricher'\n",
    "num_min_degs_for_ptw_enr=3\n",
    "tolerance_pathway_index = 0.15\n",
    "\n",
    "case_list = ['g2a_male', 'g2a_female', \n",
    "             'g2b_male', 'g2b_female', \n",
    "             'g3_male_adult',   'g3_male_elder',\n",
    "             'g3_female_adult', 'g3_female_elder']\n",
    "\n",
    "cfg = Config(project, s_project, case_list, root0)\n",
    "\n",
    "case = case_list[0]\n",
    "\n",
    "n_genes_annot_ptw, n_degs, n_degs_in_ptw, n_degs_not_in_ptw, degs_in_all_ratio = -1,-1,-1,-1,-1\n",
    "abs_lfc_cutoff, fdr_lfc_cutoff, n_degs, n_degs_up, n_degs_dw = cfg.get_best_lfc_cutoff(case, 'not_normalized')\n",
    "\n",
    "pval_pathway_cutoff = 0.05\n",
    "fdr_pathway_cutoff = .05\n",
    "num_of_genes_cutoff = 3\n",
    "\n",
    "print(f\"G/P LFC cutoffs: lfc={abs_lfc_cutoff:.3f}; fdr={fdr_lfc_cutoff:.3f}\")\n",
    "print(f\"Pathway cutoffs: pval={pval_pathway_cutoff:.3f}; fdr={fdr_pathway_cutoff:.3f}; num of genes={num_of_genes_cutoff}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start opening tables ....\n",
      "Building synonym dictionary ...\n",
      "\n",
      "\n",
      "Echo Parameters:\n",
      "For case 'g3_male_elder', there are 140/140 DAPs/DAPs with ensembl_id\n",
      "DAP's cutoffs: abs(LFC)=0.600; FDR=0.400\n",
      "\t140/140 DAPs/ensembl.\n",
      "\t\tUp 40/40 DAPs/ensembl.\n",
      "\t\tDw 100/100 DAPs/ensembl.\n",
      "\n",
      "Found 60 (best=60) pathways for geneset num=0 'Reactome_2022'\n",
      "Pathway cutoffs p-value=0.050 fdr=0.050 min genes=3\n",
      "DAPs found in enriched pathways:\n",
      "\tThere are 140 DAPs found in pathways\n",
      "\t105 (best=105) DAPs in pathways and 35/35 DAPs/ensembl not in pathways\n",
      "\n",
      "\t34 DAPs ensembl Up in pathways\n",
      "\t6 DAPs Up ensembl not in pathways\n",
      "\n",
      "\t71 DAPs ensembl Dw in pathways\n",
      "\t29 DAPs Dw ensembl not in pathways\n"
     ]
    }
   ],
   "source": [
    "pathway_name_id = 'Hemostasis - R-HSA-109582'\n",
    "pathway_name_id = 'Regulation Of IGF Transport And Uptake By IGFBPs - R-HSA-381426'\n",
    "pathway_name_id = 'Platelet degranulate - R-HSA-114608'\n",
    "pathway_name_id = 'Platelet Activation, Signaling And Aggregation - R-HSA-76002'\n",
    "pathway_name_id = 'Integrin Cell Surface Interactions - R-HSA-216083'\n",
    "pathway_name_id = 'Neutrophil Degranulation - R-HSA-6798695'\n",
    "pathway_name_id = 'Regulation of Complement cascade - R-HSA-977606'\n",
    "pathway_name_id = 'Response To Elevated Platelet Cytosolic Ca2+ - R-HSA-76005'\n",
    "\n",
    "bpx = Biopax(gene_protein, s_omics, project, s_project, root0,\n",
    "             case_list, has_age, has_gender, clone_objects=False,\n",
    "             exp_normalization=exp_normalization, geneset_num=0, \n",
    "             num_min_degs_for_ptw_enr=num_min_degs_for_ptw_enr, \n",
    "             tolerance_pathway_index=tolerance_pathway_index, \n",
    "             s_pathw_enrichm_method = s_pathw_enrichm_method)\n",
    "\n",
    "case = case_list[5]\n",
    "\n",
    "bpx.cfg.set_default_best_lfc_cutoff(normalization, abs_lfc_cutoff=1, fdr_lfc_cutoff=0.05)\n",
    "ret, degs, degs_ensembl, dfdegs = bpx.open_case(case, verbose=False)\n",
    "print(\"\\nEcho Parameters:\")\n",
    "bpx.echo_parameters()\n",
    "\n",
    "geneset_num = bpx.geneset_num"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).\n",
      "Using a pipeline without specifying a model name and revision in production is not recommended.\n",
      "/media/flalix/d2f268d1-512d-499f-b3b3-6dad7d3fdd25/anaconda3/envs/pytorch_env/lib/python3.12/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " COVID-19 is the disease caused by the virus SARS-CoV-2. When the virus enters your body, it infects cells, primarily in the respiratory system.\n"
     ]
    }
   ],
   "source": [
    "API_KEY='AIzaSyA1ZXcSe6NP5jiIw93sUpZYb8RKK1PgYDE'\n",
    "disease='COVID-19'\n",
    "context_disease=\"COVID-19 is the disease caused by the virus SARS-CoV-2. When the virus enters your body, it infects cells, primarily in the respiratory system.\"\n",
    "\n",
    "n_sentences=5\n",
    "chosen_model_list=[1,3]\n",
    "i_dfp_list=[0,1,2,3]\n",
    "\n",
    "gem=Gemini(bpx=bpx, disease=disease, context_disease=context_disease, n_sentences=n_sentences, API_KEY=API_KEY, \n",
    "             root0=root0, i_dfp_list=i_dfp_list, chosen_model_list=chosen_model_list)\n",
    "print(\"\\n\",context_disease)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "taubate_covid19 2019/10/01 2030/12/31\n",
      "Start opening tables ....\n",
      "Building synonym dictionary ...\n",
      "\n",
      "File ../src/down_pdf_pmid.sh exists True\n"
     ]
    }
   ],
   "source": [
    "prefix = s_project\n",
    "inidate=\"2019/10/01\"\n",
    "enddate=\"2030/12/31\"\n",
    "\n",
    "print(prefix, inidate, enddate)\n",
    "\n",
    "force_query = False\n",
    "verbose_query=False\n",
    "\n",
    "sleep_entrez = [30, 90, 300]; retmax=100000,\n",
    "\n",
    "''' CAP: community-acquired pneumonia\n",
    "    MV: mechanical ventilator\n",
    "'''\n",
    "remove_synonym_list =  ['CAP', 'MV', 'MDB']\n",
    "\n",
    "pub = Pubmed(bpx, gem, email, prefix, inidate, enddate, \n",
    "             root0, remove_synonym_list=remove_synonym_list, \n",
    "             sleep_entrez = [5, 7, 10], retmax=100000,  \n",
    "             try_all_text=True, text_quote='',\n",
    "             root_colab=root_colab, dec_ncpus=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PubMed Search Selected Pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(True, '../../colaboracoes/refseq/pubmed_to_reatome_terms_table.tsv')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullname = os.path.join(pub.root_refseq, pub.fname_pubmed_reatome)\n",
    "os.path.exists(fullname), fullname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway</th>\n",
       "      <th>fdr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>R-HSA-8953854</td>\n",
       "      <td>Metabolism Of RNA</td>\n",
       "      <td>0.906000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>R-HSA-194068</td>\n",
       "      <td>Bile Acid And Bile Salt Metabolism</td>\n",
       "      <td>0.255000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R-HSA-9029569</td>\n",
       "      <td>NR1H3 And NR1H2 Regulate Gene Expression Linked To Cholesterol Transport And...</td>\n",
       "      <td>0.000128</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      pathway_id                                                                          pathway  \\\n",
       "0  R-HSA-8953854                                                                Metabolism Of RNA   \n",
       "1   R-HSA-194068                                               Bile Acid And Bile Salt Metabolism   \n",
       "2  R-HSA-9029569  NR1H3 And NR1H2 Regulate Gene Expression Linked To Cholesterol Transport And...   \n",
       "\n",
       "        fdr  \n",
       "0  0.906000  \n",
       "1  0.255000  \n",
       "2  0.000128  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "force=False\n",
    "verbose=False\n",
    "\n",
    "query_type='strong'\n",
    "N=30\n",
    "\n",
    "chosen_model=3\n",
    "gemini_model='gemini-1.5-flash'\n",
    "case_sel_list = ['g3_male_adult', 'g3_female_elder']\n",
    "case=case_sel_list[0]\n",
    "\n",
    "\n",
    "dfsel = pub.gem.open_yes_no_sampling(case=case, gemini_model=gemini_model, N=N, query_type=query_type, verbose=verbose)\n",
    "dfsel.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table opened ((227, 3)) at '../../colaboracoes/refseq/pubmed_to_reatome_terms_table.tsv'\n",
      "All pathways already have terms.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway</th>\n",
       "      <th>term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>R-HSA-114608</td>\n",
       "      <td>Platelet Degranulation</td>\n",
       "      <td>Platelet Degranulation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>R-HSA-76005</td>\n",
       "      <td>Response To Elevated Platelet Cytosolic Ca2+</td>\n",
       "      <td>Platelet Calcium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R-HSA-140877</td>\n",
       "      <td>Formation Of Fibrin Clot (Clotting Cascade)</td>\n",
       "      <td>Fibrin Clot</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     pathway_id                                       pathway                    term\n",
       "0  R-HSA-114608                        Platelet Degranulation  Platelet Degranulation\n",
       "1   R-HSA-76005  Response To Elevated Platelet Cytosolic Ca2+        Platelet Calcium\n",
       "2  R-HSA-140877   Formation Of Fibrin Clot (Clotting Cascade)             Fibrin Clot"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ptw_terms = pub.calc_reactome_terms_table(dfsel, verbose=True)\n",
    "df_ptw_terms.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfa = df_ptw_terms[  pd.isnull(df_ptw_terms.term) | (df_ptw_terms.term == '')]\n",
    "dfa.empty"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway</th>\n",
       "      <th>term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>R-HSA-114608</td>\n",
       "      <td>Platelet Degranulation</td>\n",
       "      <td>Platelet Degranulation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>R-HSA-6798695</td>\n",
       "      <td>Neutrophil Degranulation</td>\n",
       "      <td>Neutrophil Degranulation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>R-HSA-5663202</td>\n",
       "      <td>Diseases Of Signal Transduction By Growth Factor Receptors And Second Messen...</td>\n",
       "      <td>Growth Factor Receptors And Second Messenger</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       pathway_id  \\\n",
       "0    R-HSA-114608   \n",
       "31  R-HSA-6798695   \n",
       "36  R-HSA-5663202   \n",
       "\n",
       "                                                                            pathway  \\\n",
       "0                                                            Platelet Degranulation   \n",
       "31                                                         Neutrophil Degranulation   \n",
       "36  Diseases Of Signal Transduction By Growth Factor Receptors And Second Messen...   \n",
       "\n",
       "                                            term  \n",
       "0                         Platelet Degranulation  \n",
       "31                      Neutrophil Degranulation  \n",
       "36  Growth Factor Receptors And Second Messenger  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ptw_pubmed = df_ptw_terms[df_ptw_terms.pathway_id.isin(dfsel.pathway_id)].copy()\n",
    "df_ptw_pubmed.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>> g3_male_adult True\n",
      "# Searching in PubMed crossing terms and pathway terms; no symbols:\n",
      " 0 0 0 0 3221 - found for for R-HSA-1643685 'Disease': AND '['OR', 'COVID', 'SARS-CoV-2']' '['male', 'severe']' '['Disease']'\n",
      " 3 - found for for R-HSA-373076 'Class A/1 (Rhodopsin-like Receptors)': AND '['OR', 'COVID', 'SARS-CoV-2']' '['male', 'severe']' '['Class', 'A/1']'\n",
      " 0 0 0 0 0 2 - found for for R-HSA-112316 'Neuronal System': AND '['OR', 'COVID', 'SARS-CoV-2']' '['male', 'severe']' '['Neuronal', 'System']'\n",
      " 0 0 0 0 0 0 2 - found for for R-HSA-8953854 'Metabolism Of RNA': AND '['OR', 'COVID', 'SARS-CoV-2']' '['male', 'severe']' '['Metabolism', 'RNA']'\n",
      " 0 0 0 7 - found for for R-HSA-1280218 'Adaptive Immune System': AND '['OR', 'COVID', 'SARS-CoV-2']' '['male', 'severe']' '['Adaptive', 'Immune', 'System']'\n",
      " 0 134 - found for for R-HSA-74160 'Gene Expression (Transcription)': AND '['OR', 'COVID', 'SARS-CoV-2']' '['male', 'severe']' '['Transcription']'\n",
      " 1 - found for for R-HSA-196071 'Metabolism Of Steroid Hormones': AND '['OR', 'COVID', 'SARS-CoV-2']' '['male', 'severe']' '['Metabolism', 'Steroid', 'Hormone']'\n",
      " 8 - found for for R-HSA-212436 'Generic Transcription Pathway': AND '['OR', 'COVID', 'SARS-CoV-2']' '['male', 'severe']' '['Transcription', 'Pathway']'\n",
      " 0 0 0------------------ end all pathways  ----------------------\n",
      "\n",
      ">>> Total pmid's: 3378\n",
      ">>> g3_male_adult False\n",
      "# Searching in PubMed crossing terms and pathway terms; no symbols:\n",
      " 13 - found for for R-HSA-114608 'Platelet Degranulation': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['Platelet', 'Degranulation']'\n",
      " 25 - found for for R-HSA-6798695 'Neutrophil Degranulation': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['Neutrophil', 'Degranulation']'\n",
      " 0 40 - found for for R-HSA-983169 'Class I MHC Mediated Antigen Processing And Presentation': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['MHC-I']'\n",
      " 9999 - found for for R-HSA-1643685 'Disease': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['Disease']'\n",
      " 21 - found for for R-HSA-373076 'Class A/1 (Rhodopsin-like Receptors)': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['Class', 'A/1']'\n",
      " 18 - found for for R-HSA-71387 'Metabolism Of Carbohydrates': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['Metabolism', 'Carbohydrate']'\n",
      " 1 - found for for R-HSA-8964058 'HDL Remodeling': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['HDL', 'Remodeling']'\n",
      " 0 0 0 120 - found for for R-HSA-112316 'Neuronal System': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['Neuronal', 'System']'\n",
      " 5 - found for for R-HSA-372790 'Signaling By GPCR': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['Signaling', 'GPCR']'\n",
      " 0 0 2 - found for for R-HSA-9013149 'RAC1 GTPase Cycle': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['RAC1']'\n",
      " 2 - found for for R-HSA-194315 'Signaling By Rho GTPases': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['Rho', 'GTPase']'\n",
      " 1 - found for for R-HSA-500792 'GPCR Ligand Binding': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['GPCR', 'Ligand']'\n",
      " 87 - found for for R-HSA-8953854 'Metabolism Of RNA': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['Metabolism', 'RNA']'\n",
      " 7 - found for for R-HSA-194068 'Bile Acid And Bile Salt Metabolism': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['Bile', 'Metabolism']'\n",
      " 3 - found for for R-HSA-389356 'CD28 Co-Stimulation': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['CD28', 'stimulation']'\n",
      " 39 - found for for R-HSA-2262752 'Cellular Responses To Stress': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['Cellular', 'Responses', 'Stress']'\n",
      " 306 - found for for R-HSA-1280218 'Adaptive Immune System': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['Adaptive', 'Immune', 'System']'\n",
      " 1 - found for for R-HSA-3299685 'Detoxification Of Reactive Oxygen Species': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['Detoxification', 'ROS']'\n",
      " 2601 - found for for R-HSA-74160 'Gene Expression (Transcription)': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['Transcription']'\n",
      " 5 - found for for R-HSA-196071 'Metabolism Of Steroid Hormones': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['Metabolism', 'Steroid', 'Hormone']'\n",
      " 341 - found for for R-HSA-212436 'Generic Transcription Pathway': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['Transcription', 'Pathway']'\n",
      " 2 - found for for R-HSA-1442490 'Collagen Degradation': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['Collagen', 'Degradation']'\n",
      " 3 - found for for R-HSA-192105 'Synthesis Of Bile Acids And Bile Salts': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['Synthesis', 'Bile']'\n",
      " 1 - found for for R-HSA-114604 'GPVI-mediated Activation Cascade': AND '['OR', 'COVID', 'SARS-CoV-2']' '['severe']' '['GPVI']'\n",
      "------------------ end all pathways  ----------------------\n",
      "\n",
      ">>> Total pmid's: 13643\n",
      "\n",
      ">>> g3_female_elder True\n",
      "# Searching in PubMed crossing terms and pathway terms; no symbols:\n",
      " 0 0 0 0 2 - found for for R-HSA-168249 'Innate Immune System': AND '['OR', 'COVID', 'SARS-CoV-2']' '['female', 'elder', 'severe']' '['Innate', 'Immune']'\n",
      " 0 0 0 1 - found for for R-HSA-449147 'Signaling By Interleukins': AND '['OR', 'COVID', 'SARS-CoV-2']' '['female', 'elder', 'severe']' '['Interleukins']'\n",
      " 1 - found for for R-HSA-1280215 'Cytokine Signaling In Immune System': AND '['OR', 'COVID', 'SARS-CoV-2']' '['female', 'elder', 'severe']' '['Cytokine', 'Signaling']'\n",
      " 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0------------------ end all pathways  ----------------------\n",
      "\n",
      ">>> Total pmid's: 4\n",
      ">>> g3_female_elder False\n",
      "# Searching in PubMed crossing terms and pathway terms; no symbols:\n",
      " 0 1 - found for for R-HSA-140877 'Formation Of Fibrin Clot (Clotting Cascade)': AND '['OR', 'COVID', 'SARS-CoV-2']' '['elder', 'severe']' '['Fibrin', 'Clot']'\n",
      " 0 1 - found for for R-HSA-76009 'Platelet Aggregation (Plug Formation)': AND '['OR', 'COVID', 'SARS-CoV-2']' '['elder', 'severe']' '['Platelet', 'Aggregation']'\n",
      " 40 - found for for R-HSA-168249 'Innate Immune System': AND '['OR', 'COVID', 'SARS-CoV-2']' '['elder', 'severe']' '['Innate', 'Immune']'\n",
      " 0 0 7 - found for for R-HSA-375276 'Peptide Ligand-Binding Receptors': AND '['OR', 'COVID', 'SARS-CoV-2']' '['elder', 'severe']' '['Peptide', 'Receptor']'\n",
      " 3 - found for for R-HSA-449147 'Signaling By Interleukins': AND '['OR', 'COVID', 'SARS-CoV-2']' '['elder', 'severe']' '['Interleukins']'\n",
      " 11 - found for for R-HSA-1280215 'Cytokine Signaling In Immune System': AND '['OR', 'COVID', 'SARS-CoV-2']' '['elder', 'severe']' '['Cytokine', 'Signaling']'\n",
      " 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 3 - found for for R-HSA-9675108 'Nervous System Development': AND '['OR', 'COVID', 'SARS-CoV-2']' '['elder', 'severe']' '['Nervous', 'Development']'\n",
      " 0 0 0------------------ end all pathways  ----------------------\n",
      "\n",
      ">>> Total pmid's: 66\n",
      "\n",
      "-------------- end --------------\n"
     ]
    }
   ],
   "source": [
    "test=False\n",
    "force=False\n",
    "verbose=False\n",
    "\n",
    "case_sel_list = ['g3_male_adult', 'g3_female_elder']\n",
    "run_list=['run01', 'run02']\n",
    "\n",
    "query_type='strong'\n",
    "N=30\n",
    "\n",
    "chosen_model=3\n",
    "gemini_model='gemini-1.5-flash'\n",
    "\n",
    "i_dfp=0\n",
    "\n",
    "for case in case_sel_list:\n",
    "    for with_gender in [True, False]:\n",
    "        print(\">>>\",  case, with_gender)\n",
    "\n",
    "        terms_not_param = ['NOT', 'MERS', 'SARS-CoV-1']\n",
    "        terms1_param = [\"OR\", 'COVID', 'SARS-CoV-2']\n",
    "        connective_param = 'AND'\n",
    "\n",
    "    \n",
    "        _ = pub.run_case_pathway_pubmed_search(case=case, i_dfp=i_dfp, with_gender=with_gender, \n",
    "                                               gemini_model=gemini_model, N=N, query_type=query_type,\n",
    "                                               force=force, verbose=verbose)\n",
    "\n",
    "    print(\"\")\n",
    "print(\"-------------- end --------------\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
